autoprotocol/protocol.py

```diff
@@ -1035,102 +1046,253 @@ def liquid_handle_dispense(
-            predispense = Unit(0, "uL")
+            len_vols = len(volume)
+            if len_vols != len_dests:
+                raise ValueError(
+                    f"When specifying more than one volume "
+                    f"the number of volumes must match the number of destination locations "
+                    f"num of volumes {len_vols} != num of destinations {len_dests}"
+                )
+            volume: List[Unit] = [
+                [parse_unit(v, "uL") for v in [vol] * len(destination[i])]
+                for i, vol in enumerate(volume)
+            ]
+
+        if not isinstance(liquid, list):
+            lc = _validate_as_instance(liquid, LiquidClass)
+            liquid: List[LiquidClass] = [lc] * len_src
         else:
-            predispense = method.predispense
-        predispense_volume = parse_unit(predispense, "uL")
+            len_liquid = len(liquid)
+            if len_liquid != len_src:
+                raise ValueError(
+                    f"The number of LiquidClass(s) {len_liquid} and sources {len_src} do not match "
+                )
+            elif not all(_validate_as_instance(lc, LiquidClass) for lc in liquid):
+                raise ValueError(
+                    f"liquid list {liquid} did not contain all of type LiquidClass"
+                )
 
-        auxiliary_locations = [
-            LiquidHandle.builders.location(
-                location=source,
-                transports=method._aspirate_transports(
-                    sum(volume) + prime_volume + predispense_volume
-                ),
-            ),
-            LiquidHandle.builders.location(
-                location=source, transports=method._prime_transports(prime_volume)
-            )
-            if method.prime
-            else [],
-            LiquidHandle.builders.location(
-                location=None,
-                transports=method._predispense_transports(predispense_volume),
+        if not isinstance(method, list):
+            dm = _validate_as_instance(method, DispenseMethod)
+            method: List[DispenseMethod] = [dm] * len_src
+        else:
+            len_method = len(method)
+            if len_method != len_src:
+                raise ValueError(
+                    f"The number of DispenseMethod(s) {len_method} and sources {len_src} do not match "
+                )
+            elif not all(_validate_as_instance(dm, DispenseMethod) for dm in method):
+                raise ValueError(
+                    f"method list {method} did not contain all of type DispenseMethod"
+                )
+
+        # Set the DispenseMethod's liquid class
+        for dm, lc in zip(method, liquid):
+            # The index of the method and liquid classes should line up with
+            # source, destination, and volumes you wish to effect. The name of
+            # the liquid class will be applied to the transports of the mappings.
+            if not dm._liquid:
+                dm._liquid = lc
```

Comment:
what are we protecting by writing it this way vs just setting `dm._liquid` without checking if there's already a liquid class?